### PR TITLE
Add support for non correlated EXISTS subquery

### PR DESCRIPTION
**Pull request by kokosing**
Add support for non correlated EXISTS subquery

### CODE REVIEW

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 1]
```diff
@@ -273,6 +272,7 @@ primaryExpression
     | identifier '->' expression                                                     #lambda
     | '(' identifier (',' identifier)* ')' '->' expression                           #lambda
     | '(' query ')'                                                                  #subqueryExpression
+    | EXISTS '(' query ')'                                                           #exists
```
Comment (martint) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: What's an example of an expression that doesn't work? Can you add tests to TestSqlParser?

File: presto-main/src/main/java/com/facebook/presto/operator/ExistsOperator.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slices;
+
+import java.util.List;
+
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class ExistsOperator
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/ExistsOperator.java]: Why do we need a physical operator for Exists? We should be able to model it as:

```
Filter[$0 > 0]
  Aggregation[$0 = count(*)]
    Limit[1]
      ...
```

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 (hunk 1 above)
Comment (kokosing) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: For example `SELECT EXISTS(SELECT true) = EXISTS(SELECT true)` didn't work.

This queries are already added to `AbstractTestQueries`. Would you like to have additional tests for them in `TestSqlParser`?

File: presto-main/src/main/java/com/facebook/presto/operator/ExistsOperator.java (hunk 2 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/ExistsOperator.java]: It would complicate things a bit. Take look at `AddExchanges#visitJoin`, there is a check if a right side of join is scalar subquery or exists subquery if so then a broadcast join is used. Having this modeled as you suggested would make this check to be complicated and vulnerable to bugs. 

But from the other side that model would enable broadcast join to be applied to broader space of queries. I will do that.

**Comment by kokosing** [comment]
@martint comments applied

**Comment by kokosing** [comment]
@martint  ping

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 (hunk 1 above)
Comment (martint) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: Interesting. If I remember correctly, that's not valid ANSI SQL, but let me take a closer look at the spec.

**Comment by martint** [comment]
I merged some of the cleanup commits. I'm looking at the rest. 

There are merge conflicts when I try to rebase. Can you take a look?

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 (hunk 1 above)
Comment (martint) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: According to the spec, that's not a valid expression. `EXISTS` is a `<predicate>`, which is a form of `<boolean primary>`, which can only be operated on with `AND`, `OR`, `NOT`, `IS/IS NOT TRUE/FALSE/UNKNOWN`.

We may want to extend Presto to support it, but we need to consider the implications (possibly grammar ambiguities, etc)

Comment (kokosing) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: I am going to extract this as a separate pull request

**Comment by kokosing** [comment]
@martint Rebased and applied comments.

**Comment by martint** [comment]
Merged, thanks!
